### PR TITLE
feat: support wildcard apps over tunnel

### LINKED ISSUE #4517: Add wildcard app support to tunnel
If I'm using the tunnel, it'd be awesome if my wildcard apps could be tunneled too. Perhaps a single-level domain could be used to avoid generating additional certificates?

```sh
# Coder control plane: 
https://fccabeef4dc13617d5ec0c06ad7ac83c.pit-1.try.coder.app

# A wildcard app:
https://34469--dev--ben-stable--bpmct--fccabeef4dc13617d5ec0c06ad7ac83c.pit-1.try.coder.app
```

### PR DESCRIPTION
**Pull request by deansheather**
Closes #4517 

Added wildcard support to the tunnel server in coder/wgtunnel#2.

If `--access-url` is undefined, a persistent tunnel URL will be generated for the server. If `--wildcard-access-url` is also not defined, it will set the wildcard access URL to `*--$TUNNEL_HOST` i.e. `*--abcdef.tunnel.dev`. Subdomain-based apps and port-forwarding should work out of the box for tunneled coder deployments.

image

### CODE REVIEW

File: cli/server.go [hunk 1]
```diff
@@ -207,15 +207,25 @@ func Server(dflags *codersdk.DeploymentFlags, newAPI func(context.Context, *code
...
 				dflags.AccessURL.Value = tunnel.URL
+
+				if dflags.WildcardAccessURL.Value == "" {
+					u, err := parseURL(ctx, tunnel.URL)
+					if err != nil {
+						return xerrors.Errorf("parse tunnel url: %w", err)
+					}
+
+					// Suffixed wildcard access URL.
+					dflags.WildcardAccessURL.Value = fmt.Sprintf("*--%s", u.Hostname())
```
Comment (coadler) [inline cli/server.go:227]: Shouldn't this just be one hypen?

Comment (coadler) [reply cli/server.go:227]: acutaly looks like ti deosnt matter

Comment (deansheather) [reply cli/server.go:227]: yeah it doesn't matter on the server side how many hyphens you use, but I used two here so it matches the two hyphens between each app component

**Review by coadler — APPROVED** [review]

**Comment by deansheather** [comment]
I'm going to merge this even though I'm still making improvements on the tunnel server side to allow for shorter hostnames. Right now, since the tunnel hostname takes up 32 characters users only have 22 characters (excluding the 8 hyphens) for their app name, agent name, workspace name and username.

I'm hoping to reduce the hostname size on tunnels from 32 characters to 13 characters, and looking into whether we can remove the agent name from the URL for named apps if we make app slugs unique per workspace.

**Pull request merged**